### PR TITLE
fabtests/mr_test: add hmem support and a new test case

### PR DESCRIPTION
**Pull request by wzamazon**
Two commits that added support for HMEM, and a test case that free and allocate buffer again.

### CODE REVIEW

**Comment by shijin-aws** [comment]
is this test being run in fabtests/pytest/prov/efa ? Seems not. Can we add it?

File: fabtests/unit/mr_test.c [hunk 1]
```diff
@@ -188,16 +206,73 @@ static int mr_regattr()
...
+			goto out;
+		}
+
+		ret = fi_close(&mr->fid);
+		if (ret) {
+			FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
+			goto out;
+		}
+
+		ret = ft_hmem_alloc(opts.iface, opts.device, (void **)&buf2, buf_size);
```
Comment (shijin-aws) [inline fabtests/unit/mr_test.c:245]: I guess we allocate a buf2 and make buf = buf2 is to make sure the new alloc has different region with the existing one. Right?

File: fabtests/unit/mr_test.c [hunk 2]
```diff
@@ -252,11 +279,9 @@ int main(int argc, char **argv)
 	if (ret)
 		goto out;
 
-	buf = malloc(test_size[test_cnt - 1].size);
-	if (!buf) {
-		ret = -FI_ENOMEM;
+	ret = ft_hmem_alloc(opts.iface, opts.device, (void **)&buf, test_size[test_cnt - 1].size);
```
Comment (shefty) [inline fabtests/unit/mr_test.c:338]: check line length and spacing

File: fabtests/unit/mr_test.c [hunk 3]
```diff
@@ -188,16 +206,73 @@ static int mr_regattr()
...
+	attr.access = FI_SEND;
+	attr.offset = 0;
+	attr.requested_key = FT_MR_KEY;
+	attr.context = NULL;
+	attr.iface = opts.iface;
+	if (opts.iface == FI_HMEM_CUDA)
+		attr.device.cuda = opts.device;
+
+	testret = FAIL;
+	for (i=0; i < 5; ++i) {
```
Comment (tschuett) [inline fabtests/unit/mr_test.c:232]: 5 is a magic number? Could it be a named constant instead?

File: fabtests/unit/mr_test.c (hunk 1 above)
Comment (wzamazon) [reply fabtests/unit/mr_test.c:245]: yes

File: fabtests/unit/mr_test.c (hunk 2 above)
Comment (wzamazon) [reply fabtests/unit/mr_test.c]: fixed line length

File: fabtests/unit/mr_test.c (hunk 3 above)
Comment (wzamazon) [reply fabtests/unit/mr_test.c]: added a variable `numtry`

**Review by shijin-aws — APPROVED** [review]

**Comment by shijin-aws** [comment]
> is this test being run in fabtests/pytest/efa ? Seems not. Can we add it?

Bump comment

**Comment by wzamazon** [comment]
> is this test being run in fabtests/pytest/efa ? Seems not. Can we add it?

It is currently run by `fabtests/pytest/efa` as part of `test_from_default.py` 

https://github.com/ofiwg/libfabric/blob/293205b9d426593c502e6207bb48029a53cf1f4d/fabtests/pytest/efa/test_from_default.py#L5

I do plan to extend it to cover GPU memory

File: fabtests/common/shared.c [hunk 4]
```diff
@@ -381,6 +407,7 @@ int ft_reg_mr(struct fi_info *fi, void *buf, size_t size, uint64_t access,
 	int ret;
 	uint64_t flags;
 
+
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply fabtests/common/shared.c]: Thank you!

fixed in new revision.

**Comment by wzamazon** [comment]
@shefty Can you help with the "continuous-integration/jenkins/pr-head" failure?

Also, can you please review again?

**Comment by wzamazon** [comment]
@shefty 

Can you help me understand the "continuous-integration/jenkins/pr-head" failure ?

Also I believe I addressed you earlier comments. Can you review again?

**Comment by aingerson** [comment]
@wzamazon Jenkins failure is a transient error we've been seeing lately and working to resolve. Unrelated, you can ignore.

**Review by shefty — APPROVED** [review]

**Comment by wzamazon** [comment]
@aingerson @shefty 

I will merge then

**Pull request merged**